Always draw `disk_service` routes

stefannibrasil opened the pull request:
### Summary

Fixes #42043 

When `config.active_storage.draw_routes = false` while the current storage adapter is using the `disk_service`, and one attempts to call the url method on an attachment (`model.attachment.url`), it raises this error:
```
Active Storage Disk Service NoMethodError: undefined method `rails_disk_service_url' 
``` 

This PR fixes this issue by doing the following:
- Always draw ActiveStorage routes when the current storage adapter is using `disk_service`.

### Bug fix reproducing steps

A gist with the reproducing steps appointing to our branch can be found [here](https://gist.github.com/stefannibrasil/e8050ad97ab32e14f07dca651e186703).

lfalcao commented:
@stefannibrasil why split these routes in a different file?

stefannibrasil commented:
Hi @lfalcao it was suggested [here](https://github.com/rails/rails/issues/42043#issuecomment-835718249).

lfalcao reviewed (commented):
> Hi @lfalcao it was suggested [here](https://github.com/rails/rails/issues/42043#issuecomment-835718249).

Ok, well you have to somehow add this new route file path (`activestorage/config/routes/active_storage.rb`) into draw_paths so `draw` method can load it correctly. It's [crashing because the CI](https://buildkite.com/rails/rails/builds/78063#9a1dcfb7-34a2-4efc-bb37-c215129a7b9a) is trying to find `routes/active_storage.rb` inside the app `config/routes` path, not inside the gem dir... :thinking: 

I'm lazy, so I would make this in just one file. :blush: 

```ruby
Rails.application.routes.draw do
  if ActiveStorage.draw_routes
    # all routes, except those bellow
  end

  # explain why this exists
  if ActiveStorage.disk_service_enabled? || ActiveStorage.draw_routes
    # disk service routes
  end
end
```

stefannibrasil commented on activestorage/lib/active_storage.rb:
```diff
@@ -68,6 +68,11 @@ module ActiveStorage
   mattr_accessor :track_variants, default: false
 
+  def self.disk_service_enabled?
+    local_service = Blob.services.fetch(:local)
```
That's a great point, thanks for pointing that out @lfalcao We'll double check that 👍 

Regarding the files not being loaded, we tried a few things with no success. We'll try a bit more, otherwise we'll move everything back to the same file.

Thank you for the tips, that's super helpful!

zzak commented:
@stefannibrasil There's a bunch of failures in CI, though I'm not sure how related they are. For good luck, could you rebase and squash your commits? 🙏

sbs-cats committed:
Always draw ActiveStorage routes for disk_service

Always draw routes for ActiveStorage when the adapter
being used is a DiskService, or a MirrorService
with at least one DiskService enabled.

stefannibrasil commented:
hi @zzak thank you for your support. We had a few errors after merging with main, they should be fixed now.

stefannibrasil commented:
@lfalcao @pixeltrix this PR is finally ready for review. Thank you in advance for the guidance :100:

stefannibrasil commented:
I saw this [open PR](https://github.com/rails/rails/pull/42593) and I wanted to ask if you suggest us updating the docs here to make it explicit that the local disk service url routes will be generated even if `draw_routes = false`?

pixeltrix commented:
> I saw this [open PR](https://github.com/rails/rails/pull/42593) and I wanted to ask if you suggest us updating the docs here to make it explicit that the local disk service url routes will be generated even if `draw_routes = false`?

Yes, that's probably a good idea - maybe point that the disk service is intended for development and test use and not production.

pixeltrix commented on activestorage/lib/active_storage.rb:
```diff
@@ -71,6 +71,10 @@ module ActiveStorage
 
+  def self.disk_service_enabled?
+    Blob.service&.is_a? Service::DiskService
+  end
```
What happens if there's a mirror service configured and one of the mirrors is a disk service?

stefannibrasil replied:
Good point! We can probably handle this case by checking if the service is a mirror and if any of its mirrors/primary service is `Disk::Service`.

guilleiguaran replied:
@stefannibrasil hi! can you implement your suggestion?

stefannibrasil replied:
hi @guilleiguaran yes, we're working on it.

thdaraujo replied:
should be good to review now! 🙌

zzak commented:
Ok, so IMO `draw_routes=false` is doing exactly what it was intended to do.

What is the purpose of turning off those routes but still want the URL helpers? You cannot have one without the other.

Perhaps if you just wanted _some_ of the routes but not all of them we could provide that feature (draw_routes could take some params).. but I'd like to know a bit more what you're trying to do. 🤔

stefannibrasil commented:
sorry, closed PR by mistake after squashing commits.

Hopefully this will fox all the issues :pray:

jmarsh24 commented:
Why was this PR closed? It's still an issue with the rails proxy mode that needs to be addressed?
